# Metabolome permafrost

This is a repository containing the code for replicating the analysis and the figures for the paper: 

**Microbial-Metabolite Dynamics Illuminate Permafrost Ecosystem Complexity at the Feature Scale**

## Assembly Microbiome

1. **`R scripts from 1 to 2`** were used to calculate the ecological processes contributing to microbiome assembly. The input of these scripts include a phylogenetic tree derived from metagenome assembled geneomes and their abundances. Author Hannah Holland Moritz

## Assembly Metabolome

1. **`R scripts from 1 to 5`** were adapted from Danczak et al. 2020, https://github.com/danczakre/Meta-Metabolome_Ecology to calculate the assembly processes contributing to the peat metabolome. The input of the first script is a report of the relative abundances of masses detected through FT-ICR MS. The outputs of each script serve as input of the following ones.

## Assembly at feature level: microbiome and metabolome

1. **`R scripts in these sections`** were adapted from Danczak et al. 2022, https://github.com/danczakre/betaNTI-feature to calculate the &beta;NTI<sub>feature</sub> value within each habitat across the permafrost gradient. This analysis were performed for the microbiome and metabolome separately. The input for the microbiome analysis include a phylogenetic tree derived from metagenome assembled geneomes and their abundances. The input for the metabolome analysis include the dedrogram (TWCD) and processed report derived from the `Assembly Metabolome` section.

##  &beta;NTI<sub>feature</sub> clustering

1. **`Cluster_bnti_features.Rmd`**: Script for clustering metabolites based on their &beta;NTI<sub>feature</sub> value and identifying the representative features per each cluster. Inputs are &beta;NTI<sub>feature</sub> matrices filtered by habitat. Outputs are list of representative features per habitat (`rep_features_{habitat}.csv` files) and the  &beta;NTI<sub>feature</sub> values of the representative features (`rep_features_{habitat}_bnti.csv` files).

1. **`Cluster_correlations.Rmd`**: Script to calculate the Spearman correlations between the consensus &beta;NTI<sub>feature</sub> value of each cluster in each sample against microbial &beta;NTI<sub>feature</sub> values, microbial abundances, environmental data (i.e., depth, soil temperature, %C %N, and precipitation) and greenhouse gas (CO<sub>2</sub> and CH<sub>4</sub>) data. Inputs are the the lists of representative features per habitat generated by the `Cluster_bnti_features.Rmd` and csv files of the previously mentioned datasets. Outputs are csv files with the results of the correlations with every data type (`*_correlations.csv` files).

## Figures

1. **`Figure_1.Rmd`**: Script to generate the complete **Figure 1** of the paper, including the comparison of the average within-habitat &beta nearest taxon index (βNTI) and the ecological processes driving the microbiome and metabolome across the permafrost gradient. Inputs for this figure derive from results of the `Assembly Microbiome` and `Assembly Metabolome` folders. Also, it includes correlations of the beta nearest taxon index (βNTI) with environmental factors. Output is Figure 1.

1. **`Figure_2.Rmd`**: Script to generate the complete **Figure 2** of the paper, including the correlation networks of the &beta;NTI<sub>feature</sub> clusters, as well as the biogeochemical indexes, metabolomome assembly contribution, and elemental composition of the representative features. Inputs are the correlation results generated by the `Cluster_correlations.Rmd` script, and the &beta;NTI<sub>feature</sub> values of the representative features generated by the `Cluster_bnti_features.Rmd` script. Output is Figure 2.

1. **`Figure_3.Rmd`**: Script to generate the complete **Figure 3** of the paper, including the correlation to greenhouse gases (GHG) and functional potential of bacteria associated with the &beta;NTI<sub>feature</sub> clusters. Inputs are abundances, &beta;NTI<sub>feature</sub> values, and metagenome annotation tables of bacteria (performed with [METABOLIC](https://github.com/AnantharamanLab/METABOLIC)) from the bog as well as GHG concentration data. Output is Figure 3.

1. **`Figure_4A-C_Supplementary_4.Rmd`**: Script to generate the first half of **Figure 4** of the paper, the microbial-metabolite co-occurrence networks within each habitat. Inputs are the network's edge files produced by [MENAP](http://ieg4.rccc.ou.edu/mena/) and chemical information (e.g., elemental composition) of the metabolites in the networks (`node_table_annotated.csv`) derived from previous analysis. Additionally this scripts generates **Supplementary Figure 4**, which includes information about the networks (e.g., number of nodes and edges). Outputs are panels A through C of  Figure 4, Supplementary Figure 4, and a table detailing the bacteria present in the networks (`networked_bacteria.csv`).

1. **`Figure_4D-F_Supplementary_5.Rmd`**: Script to generate the other half of **Figure 4** of the paper, the Sankey diagrams of the networked bacteria. Inputs are the taxonomical annotation of the networked bacteria (`network_nodes_taxonomy.csv`), the phylogenetic tree of the bacteria (provided by Hannah Holland-Moritz) and the Sankey diagram inputs and annotation tables produced by [METABOLIC](https://github.com/AnantharamanLab/METABOLIC) (`Metabolic_Sankey_diagram_input_*.txt` files). Additionally this scripts generates **Supplementary Figure 4**, which indicates the number of annotated CAZymes present in the networked bacteria. Output is panels D through F of Figure 4 and Supplementary Figure 5.

1. **`Supplementary_1.R`**: Script to generate **Supplementary Figure 1** of the paper which includes the &beta;NTI<sub>feature</sub> contribution and the relative percentage of metabolite elemental composition within each habitat. The relative abundance of metabolite elemental compositions and the relative abundance of different metabolite classes in the &beta;NTI<sub>feature</sub> devided clusters. Output is Supplementary figure 1.

1. **`Supplementary_2.Rmd`**: Script to generate **Supplementary Figure 2** of the paper which includes the nitrogen and sulfur functional potential of metagenome assembled genomes (MAGs) that correlated with &beta;NTI~feature~-derived metabolite clusters. Inputs include bacterial abundances and &beta;NTI<sub>feature</sub> feature values of microbes that correlated with &beta;NTI<sub>feature</sub>-derived clusters as determined by the `Cluster_correlations.Rmd` script, and annotation tables produced by [METABOLIC](https://github.com/AnantharamanLab/METABOLIC) for this bacteria. Output is Supplemenatary Figure 2.

1. **`Supplementary_3.Rmd`**: Script to generate **Supplementary Figure 3** of the paper which includes the functional potential of MAGs that correlated with GHGs in the bog. Inputs are annotation tables produced by [METABOLIC](https://github.com/AnantharamanLab/METABOLIC). Output is supplementary Figure 3.
